lib/memo_wise.rb

```diff
@@ -3,6 +3,49 @@
 require "memo_wise/version"
 
 module MemoWise
-  class Error < StandardError; end
-  # Your code goes here...
+  def initialize(*values)
+    @_memo = {}
+    super
+  end
+
+  def self.prepended(target)
+    class << target
+      # Implements memoization for the given method name.
+      #
+      # The method must be a *pure function* of the value attributes, and
+      # therefore referentially transparent. Replaces the method with a new
+      # method of the same name, which memoizes the result of calling the
+      # original method. See module comment above for a usage example.
+      #
+      # @param method_name [Symbol]
+      #   Name of method for which to implement memoization.
+      def memo_wise(method_name)
+        not_memoized_name = "#{method_name}_not_memoized".to_sym
+        alias_method not_memoized_name, method_name
+        private not_memoized_name
+
+        define_method(method_name) do |*args|
+          key = [method_name, args].freeze
+          @_memo.fetch(key) do
+            @_memo[key] = send(not_memoized_name, *args)
+          end
+        end
+      end
+
+      # Undoes memoization for the given method name. If the method was
+      # previously memoized, it will no longer be memoized.
+      #
+      # @param method_name [Symbol]
+      #   Name of method for which to undo memoization.
+      def reset_memo_wise(method_name)
```

Comment:
I'm not sure how to think of this. I buy the argument that we don't use it, but I'm unsure how to know or weigh if others do or would use it. Any thoughts there?

Reply:
@JacobEvelyn I think we should split-track your two points: (a) do we need the `reset_memo_wise` in this gem, given that Panorama doesn't currently use it, and (b) separately, do we need either an API or just documentation on how to pre-set a memoized value, or to un-set or re-set an alread-memoized value, because we do those things at Panorama by hacking the instance variable

Reply:
My perspective on _(a) do we need the reset_memo_wise in this gem, given that Panorama doesn't currently use it?_:

I think we **do need this in a public gem**, because I see this as expected API surface for a competitive memoization library in the public sphere to have -- even if Panorama code doesn't currently use it. In terms of trade-offs, I don't currently perceive the size of this code+specs, nor the cognitive load of the complexity of this feature in the API, to be large enough to trade-off against having this feature. In terms of Ruby DSLs in general, I find having a way to reset or undo changes to a class to be uniformly expected in the wider community, particularly with regard to methods being aliased. Thoughts?

Reply:
My current perspective on _(b) separately, do we need either an API or just documentation on how to pre-set a memoized value, or to un-set or re-set an already-memoized value, because we do those things at Panorama by hacking the instance variable currently_:

I think we **do need this feature in a public API**, because it's useful enough that we depend on hacking around it in Panorama code. But I'd be happy seeing it come in a subsequent PR after this initial one is merged. Thoughts?

Reply:
I'm happy to be overridden here, and I hear what you're saying @ms-ati. Thanks for the thoughts!

> (a) do we need the reset_memo_wise method in this gem, given that Panorama doesn't currently use it?

I guess I don't see the harm in leaving this out until either we or another gem user needs it. While I do see this functionality in other gems like `Memoist`, a large API footprint in those gems has actually been something that discourages me from using them.

> (b) separately, do we need either an API or just documentation on how to pre-set a memoized value, or to un-set or re-set an already-memoized value, because we do those things at Panorama by hacking the instance variable currently?

If we can come up with a reasonable API that we ourselves would use, I'm totally down to add it. (And I agree with you that we can wait for a subsequent PR to add it.)

Reply:
(moved to a later PR)